### ISSUE TITLE
UNLs with short file parts should use all open commanders

### ISSUE THREAD

**Issue by boltex** [leo-editor/leo-editor#3811]
**Background (EKR)**

See PR #3812. Experiments (in test.leo) show the following:

- UNLs with *empty* file parts use `@data unl-path-prefixes` and then all open commanders.
- UNLs with *short* file parts **fail** unless the short file part is in `@data unl-path-prefixes`.
- UNLs with *long (absolute)* file parts fail unless the exact give file exists.

**To do**

- UNLs with *short* file parts should work much like UNLs with *empty* file parts.

**Investigation**

- An improved unit test shows that `g.openUNLFile` works as described above.
- **Legacy** (path-based) UNLs with short file names *do* search all open commanders.
- Oops! **New** (gnx-based) UNLs with short file names search only c.

The **culprit** is `g.findAnyUnl`. The bug is here:
```python
# If there is a file part, search *only* the given commander!
if file_part:
    c2 = g.openUNLFile(c, file_part)
    if not c2:
        return None
    p = g.findGnx(tail, c2)
    return p  # May be None.
```
The proper code is:
```python
# First, search the open commander.
# #3811: Do *not* fail if this search fails.
if file_part:
    c2 = g.openUNLFile(c, file_part)
    if c2:
        p = g.findGnx(tail, c2)
        if p:
            return p
```

**Comment by tbpassin** [comment]
I definitely think that relative UNLs should resolve relative to the directory of the invoking outline.

I also proposed that if a UNL could not be found, then the recent files list should be checked to see if a outline's name could be matched there.  @Edward didn't like that, and I suppose he's probably right because the search locations wouldn't be stable over time.

I am perhaps more sensitive to most about being able to locate files because I run a lot of Linux VMs, and I'm always copying files and directories from, say, my main  Windows machine to one or another of the VMs.  Having unls (or gnxs with paths) that would keep working could be very important, and the only way that could happen easily is with relative files resolving the same way on all the VMs.

**Comment by boltex** [comment]
@edreamleo Thank you ! I'll assign you to this issue instead, and I'll focus on finishing the few issues that remain for LeoJS 1.0 and the next LeoInteg version ! :)

**Comment by edreamleo** [comment]
@boltex You're welcome :-)

**Comment by boltex** [comment]
@edreamleo Thanks for your work on this, but I see that this fixes another bug that was not my intended proposed enhancement. 

As @tbpassin mentioned above (while referencing @edward instead of @edreamleo - maybe you missed that comment) : My original intended feature to be supported/implemented, is that short UNLs should first try to open a leo file if not already opened, **not only look into already opened outlines,** to resolve relative to the directory of the invoking outline. (**As-if that was a full path UNL**)  

That is what I meant in my original title : "short" UNLs should try to resolve file-path in their commander's path as default. _(by 'resolved' I also meant 'open the outline' not just 'look into already opened outlines'.)_

But I see now that my phrasing was ambiguous and not very clear. I hope this is something you also acknowledge would be a very nice feature to have. If not, let me try to justify this with the following...

So my intention is to alter the second bullet that you've put in the opening comment of this thread:

> UNLs with short file parts fail unless the short file part is in @data unl-path-prefixes.

To me, this absolute 'forced' usage of the  @data unl-path-prefixes for short UNL to resolve and open other outlines, _even if they exist in the same folder_, is counterintuitive for new (and not-so-new) users. 😄 

So to clarify, my proposed enhancement, would be that a simple short UNL like 

unl:gnx://test2.leo#felix.20240210225359.1 residing in an outline test1.leo, existing in the same folder as test2.leo

would **first** try to open it as if a full path UNL **relative to the current directory of the invoking outline.**

So that a user who's making a project with two leo files, test1.leo and test2.leo, residing in the same project folder, could use short UNL's without having to know about and modify configuration settings. 

**Some use cases**: 
 - someone could 'zip' its project to give it to someone else,
 - or make a repository on github, and someone else clones it,
  
 ...and in either cases, the short UNL's in both test1.leo and test2.leo would work to open and find nodes in each other, without requiring the use of full-path-UNLs, nor the usage of adding @data unl-path-prefixes setting in myleosetting or an @setting node.

To me, this feels like a very natural and simple behavior that I would expect the short UNLs to have. 

Please correct me if I'm misunderstanding an aspect of UNLs or if I'm confused on the importance of forcing the @data unl-path-prefixes setting at all times for short UNLs, even if the file path could be resolved relative to the directory of the invoking outline. (as @tbpassin puts it)

Again, thanks for your time on this issue! And thanks for considering the explanations I tried to lay out more precisely 😄 (my english is sometimes not very good!)

**Comment by edreamleo** [comment]
Closed via PR #3812.

@boltex See #3814 for a discussion of another possible enhancement.
